ratelimit.c

```diff
@@ -20,9 +20,46 @@
  * SOFTWARE.
  */
 
-#include <stdlib.h>
-#include <sys/time.h>
 #include "redismodule.h"
+#include <stdlib.h>
+#include <time.h> // clock_gettime, CLOCK_MONOTONIC and CLOCK_REALTIME
+
+#ifndef CLOCK_REALTIME
+#include <sys/time.h> // gettimeofday
+#endif
+
+/* nanoseconds per second */
+#define NSEC_PER_SEC 1000000000LL
+
+/* microseconds per second */
+#define USEC_PER_SEC 1000000LL
+
+/* miliseconds per second */
+#define MSEC_PER_SEC 1000LL
+
+/* get_nanos returns the nanosecond-precise time, relative to some arbitrary
+ * point. We prefer to use the monotonic clock provided by
+ * clock_gettime(CLOCK_MONOTONIC, ...).
+ *
+ * If the monotonic clock isn't available we'll use the realtime clock from
+ * clock_gettime() or gettimeofday(). Note that these two clocks can jump
+ * forwards and backwards as the system time-of-day clock is changed, including
+ * by NTP.
+ */
+static long long get_nanos() {
+#ifndef CLOCK_REALTIME
+  struct timeval tv;
+  gettimeofday(&tv, 0);
+  return (((long long) tv.tv_sec) * NSEC_PER_SEC + tv.tv_usec) * MSEC_PER_SEC;
```

Comment:
@kleisauke I think there might be a bug here. Shouldn't we use `USEC_PER_SEC` instead of `NSEC_PER_SEC` on the first multiplication?

Reply:
I've removed the `gettimeofday()` in favor of `clock_gettime()` with https://github.com/onsigntv/redis-rate-limiter/pull/4/commits/585e76592723594c2983768709b7856bec2bfe07, so this is no longer a issue.